### ISSUE TITLE
Passing data on GPU to OpenMM via Python API and Context.setPositions

### ISSUE THREAD

**Issue by jonathanking** [openmm/openmm#3290]
I am interested in taking data I have stored on the GPU (PyTorch CUDA Tensors containing protein coordinate information) and passing this information to OpenMM to compute the energy of a system using the GPU for fast computation.

Is there any way to hand off this data without first moving the data to the CPU (numpy ndarrays)? In particular, I would like to be able to update the positions of the simulation and call `simulation.context.setPositions` on my CUDA array, rather than on a numpy ndarray. I see that there is a CudaArray object in OpenMM, but I don't know if it is possible to utilize it using Python. Simply handing off PyTorch tensors causes the kernel to crash.

I would greatly appreciate any assistance or advice you can provide.

**Comment by jchodera** [comment]
@peastman: This seems like a use case for a PyTorch `Integrator`, where the energies and forces are fed to a PyTorch model to use in computing the next set of positions. Is it possible for us to extend `openmm-torch` to do this, since this would also allow for a great deal more flexibility in implementing custom integrators?

**Comment by peastman** [comment]
That would definitely be a possibility.  We'll need to think very carefully about the design.  @jonathanking can you say more about your use case?  What does your PyTorch code do?  Is it something that could be viewed as a integrator, or is it something very different where that approach would be too restrictive?

**Comment by jonathanking** [comment]
I wouldn't describe my use case as needing an integrator. I am not developing PyTorch code to compute the next set of positions,  since my PyTorch code is not really written in the context of molecular dynamics simulations.  Instead, I only need to query the energy of a system at a single point in time, and for that, I create a single "simulation" context in OpenMM without any progressing timesteps.

I have a PyTorch machine learning model that predicts the atomic coordinates of proteins given sequence information. I also have a PyTorch loss function that takes the predicted coordinates and computes the protein's energy by passing this information to OpenMM's Python API, computing the energy, and returning the energy value back to PyTorch as a scalar loss tensor for model training. The part of my code that requires OpenMM's support really does not have anything to do with predictive modeling. However, I'm afraid that passing a large CUDA tensor of protein atomic coordinates from PyTorch to OpenMM via the CPU is going to limit the efficiency of my methods.

**Comment by peastman** [comment]
Have you measured it to see whether that's really the bottleneck?  In typical applications, the time for doing the data transfer is usually less than the time to compute the energy, though it isn't negligible.  And depending on your application, the energy evaluation might take much less time than your PyTorch model.  Before trying to optimize the data transfer, it would be good to determine how much slowdown it's really causing.

**Comment by peastman** [comment]
Here's a very simple way to measure it: try leaving out the call to `setPositions()`.  Run your code as usual, but have it evaluate the energy of the same conformation every time.  How much does that speed it up?  This gives you an upper limit on how much speedup you could possibly hope for by optimizing the data transfer.

**Comment by jonathanking** [comment]
Thank you for the suggestion. Your intuition was right–other parts of my code make up the majority of the computation time. I'll keep an eye out on this issue if the situation changes, as my code might also become less expensive in the future.

**Comment by peastman** [comment]
Thanks, that's good to know.  I'm going to close this then.  You can reopen it or create a new issue if you find a use case where it would actually make a difference.
